File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    Set<List<String>> notDuplicate = new HashSet<>();
    
    public void brute(List<Integer> seperate, String s, int index) {
        addAnswer(new ArrayList<>(seperate), s);
        
        for (int i = index + 1; i < s.length(); i++) {
            List<Integer> newSeperate = new ArrayList<>(seperate);
            newSeperate.add(i);
            brute(newSeperate, s, i);
        }
    }
    
    public void addAnswer(List<Integer> seperate, String s) {
        seperate.add(s.length());
        String[] splitS = new String[seperate.size() - 1];
        
        for (int i = 0; i < splitS.length; i++) {
            splitS[i] = s.substring(seperate.get(i), seperate.get(i + 1));
            
            if (!isPalindrome(splitS[i])) {
                return;
            }
        }
        
        List<String> addList = new ArrayList<>();
        for (int i = 0; i < splitS.length; i++) {
            addList.add(splitS[i]);
        }
        
        notDuplicate.add(addList);
    }
    
    public boolean isPalindrome(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (int i = 0; i < s.length(); i++) {
            stack.push(s.charAt(i));
        }
        
        int index = 0;
        while (!stack.isEmpty()) {
            if (stack.pop() != s.charAt(index++)) {
                return false;
            }
        }
        
        return true;
    }
    
    public List<List<String>> partition(String s) {
        List<Integer> seperate = new ArrayList<>();
        seperate.add(0);
        brute(seperate, s, 0);
        
        List<List<String>> answer = new ArrayList<>();
        for (List<String> component : notDuplicate) {
            answer.add(component);
        }
        
        return answer;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>>answer;
    bool is(string s){
        int n=s.size();
        for(int i=0;i<s.size();i++){
            if(s[i]!=s[n-i-1]){
                return false;
            }
        }
        return true;
    }
    void dfs(string& s,int idx,vector<string>a){
        if(idx+1==s.size()){
            for(string ss:a){
                if(!is(ss)){
                    return;
                }
            }
            answer.push_back(a);
            return;
        }
        for(int i=idx+1;i<s.size();i++){
            string ss=s.substr(idx+1,i-idx);
            a.push_back(ss);
            dfs(s,i,a);
            a.pop_back();
        }
    }
    vector<vector<string>> partition(string s) {
        dfs(s,-1,{});
        return answer;
    }
};